File: .agents/scratch/flatten_script.py
import os
import datetime

# Configuration
WORKSPACE_ROOT = '/Users/test/Projects/Unbox'
OUTPUT_FILE = os.path.join(WORKSPACE_ROOT, 'flatten.txt')

EXCLUDE_DIRS = {
    'node_modules', 'dist', 'build', 'target', 'bin', 'obj', 'out', 'cache',
    'vendor', 'venv', 'env', '.git', '.next', '.cache', '.agents', '.gemini',
    'lib', 'broadcast' # foundry broadcast
}

EXCLUDE_FILES = {
    'package-lock.json', 'yarn.lock', 'pnpm-lock.yaml', 'flatten.txt', '.DS_Store'
}

EXCLUDE_EXTENSIONS = {
    '.png', '.jpg', '.jpeg', '.mp4', '.pdf', '.exe', '.dll', '.class', 
    '.pyc', '.woff', '.woff2', '.ico', '.log', '.zip', '.gz', '.tar'
}

def should_exclude(path, is_dir=False):
    dirname = os.path.basename(path)
    if dirname.startswith('.') and dirname != '.':
        return True
    
    if is_dir:
        return dirname in EXCLUDE_DIRS
    
    # File checks
    ext = os.path.splitext(dirname)[1].lower()
    if ext in EXCLUDE_EXTENSIONS:
        return True
    if dirname in EXCLUDE_FILES:
        return True
    
    return False

def get_file_tree(startpath):
    tree = []
    for root, dirs, files in os.walk(startpath):
        # In-place modify dirs to skip excluded ones
        dirs[:] = [d for d in dirs if not should_exclude(os.path.join(root, d), is_dir=True)]
        
        level = root.replace(startpath, '').count(os.sep)
        indent = ' ' * 4 * (level)
        tree.append(f'{indent}{os.path.basename(root)}/')
        subindent = ' ' * 4 * (level + 1)
        for f in files:
            if not should_exclude(os.path.join(root, f)):
                tree.append(f'{subindent}{f}')
    return "\n".join(tree)

def main():
    print(f"Starting flattening at {WORKSPACE_ROOT}")
    
    all_files = []
    for root, dirs, files in os.walk(WORKSPACE_ROOT):
        # Prune dirs
        dirs[:] = [d for d in dirs if not should_exclude(os.path.join(root, d), is_dir=True)]
        
        for f in files:
            full_path = os.path.join(root, f)
            if not should_exclude(full_path):
                all_files.append(full_path)
    
    all_files.sort()
    
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f_out:
        # Header
        f_out.write(f"UNBOX CODEBASE FLATTENED\n")
        f_out.write(f"TIMESTAMP: {datetime.datetime.now().isoformat()}\n")
        f_out.write(f"WORKSPACE: {WORKSPACE_ROOT}\n")
        f_out.write(f"\n================================================================\n")
        f_out.write(f"FILE TREE\n")
        f_out.write(f"================================================================\n")
        f_out.write(get_file_tree(WORKSPACE_ROOT))
        f_out.write(f"\n\n")
        
        # Files content
        file_count = 0
        for file_path in all_files:
            rel_path = os.path.relpath(file_path, WORKSPACE_ROOT)
            try:
                with open(file_path, 'r', encoding='utf-8', errors='ignore') as f_in:
                    content = f_in.read()
                    
                f_out.write(f"================================================================\n")
                f_out.write(f"FILE: {rel_path}\n")
                f_out.write(f"================================================================\n")
                f_out.write(content)
                f_out.write(f"\n\n")
                file_count += 1
            except Exception as e:
                print(f"Skipping {rel_path} due to error: {e}")
                
    print(f"Done! Included {file_count} files.")
    size_kb = os.path.getsize(OUTPUT_FILE) / 1024
    print(f"Final size: {size_kb:.2f} KB")

if __name__ == "__main__":
    main()

File: .agents/scratch/flatten_script.js
const fs = require('fs');
const path = require('path');

const ROOT = '/Users/test/Projects/Unbox';
const OUTPUT_FILE = path.join(ROOT, 'flatten.txt');

const EXCLUDED_DIRS = [
  'node_modules', 'dist', 'build', 'target', 'bin', 'obj', 'out',
  'cache', 'broadcast', 'lib', '.git', '.next', '.cache', '.agents', '.gemini'
];

const EXCLUDED_EXTENSIONS = [
  '.png', '.jpg', '.jpeg', '.gif', '.mp4', '.pdf', '.exe', '.dll',
  '.class', '.pyc', '.woff', '.woff2', '.ttf', '.eot', '.ico', '.svg',
  '.lock', '.log'
];

const EXCLUDED_FILES = [
  'package-lock.json', 'yarn.lock', 'pnpm-lock.yaml', 'flatten.txt', '.DS_Store'
];

function getFiles(dir, allFiles = []) {
  const files = fs.readdirSync(dir);
  for (const file of files) {
    const filePath = path.join(dir, file);
    const relativePath = path.relative(ROOT, filePath);

    if (file.startsWith('.') && file !== '.env') {
        // Skip hidden unless it's .env if we want it (usually not for flattening, but follow rules)
        // Rule says All hidden directories and files (anything starting with a dot)
        continue;
    }

    if (EXCLUDED_DIRS.includes(file)) continue;
    if (EXCLUDED_FILES.includes(file)) continue;
    if (EXCLUDED_EXTENSIONS.includes(path.extname(file))) continue;

    if (fs.statSync(filePath).isDirectory()) {
      getFiles(filePath, allFiles);
    } else {
      allFiles.push(filePath);
    }
  }
  return allFiles;
}

function generateTree(dir, prefix = '') {
    let tree = '';
    const files = fs.readdirSync(dir).filter(f => {
        if (f.startsWith('.')) return false;
        if (EXCLUDED_DIRS.includes(f)) return false;
        return true;
    });

    for (let i = 0; i < files.length; i++) {
        const file = files[i];
        const filePath = path.join(dir, file);
        const isLast = i === files.length - 1;
        const line = prefix + (isLast ? '└── ' : '├── ') + file + '\n';
        tree += line;

        if (fs.statSync(filePath).isDirectory()) {
            tree += generateTree(filePath, prefix + (isLast ? '    ' : '│   '));
        }
    }
    return tree;
}

const allSourceFiles = getFiles(ROOT);
const timestamp = new Date().toISOString();
const tree = generateTree(ROOT);

let output = `================================================================
FLATTENED CODEBASE: Unbox
TIMESTAMP: ${timestamp}
================================================================

FILE TREE:
.
${tree}

`;

for (const file of allSourceFiles) {
  const relativePath = path.relative(ROOT, file);
  const content = fs.readFileSync(file, 'utf8');
  output += `================================================================
FILE: ${relativePath}
================================================================
${content}

`;
}

fs.writeFileSync(OUTPUT_FILE, output);
console.log(`Successfully created flatten.txt with ${allSourceFiles.length} files.`);
